Core/include/Acts/Utilities/detail/Subspace.hpp

```diff
@@ -0,0 +1,368 @@
+  template <typename scalar_t>
+  using ExpansionMatrix = Eigen::Matrix<scalar_t, kFullSize, kSize>;
+
+  // the functionality could also be implemented using a std::bitset where each
+  // bit corresponds to an axis in the fullspace and set bits indicate which
+  // bits make up the subspace. this would be a more compact representation but
+  // complicates the implementation since we can not easily iterate over the
+  // indices of the subspace. storing the subspace indices directly requires a
+  // bit more memory but is easier to work with. for our typical use cases with
+  // n<=8, this still takes only 64bit of memory.
+  std::array<uint8_t, kSize> m_axes;
+
+ public:
+  /// Construct from an array of axis indices.
+  ///
+  /// @tparam index_t Scalar index type
+  /// @param indices Ordered array of unique axis indices
+  template <typename index_t>
+  constexpr FixedSizeSubspace(const std::array<index_t, kSize>& indices) {
+    // verify inputs
+    for (auto i = 0u; i < kSize; ++i) {
+      assert((indices[i] < kFullSize) and
+             "Axis indices must be within the full space");
+      if ((i + 1u) < kSize) {
+        assert((indices[i] < indices[i + 1u]) and
+               "Axis indices must be strongly ordered and unique");
+      }
+    }
+    // assign stored axis
+    for (auto i = 0u; i < kSize; ++i) {
+      m_axes[i] = static_cast<uint8_t>(indices[i]);
+    }
+  }
+  // The subset can not be constructed w/o defining its axis indices.
+  FixedSizeSubspace() = delete;
+  FixedSizeSubspace(const FixedSizeSubspace&) = default;
+  FixedSizeSubspace(FixedSizeSubspace&&) = default;
+  FixedSizeSubspace& operator=(const FixedSizeSubspace&) = default;
+  FixedSizeSubspace& operator=(FixedSizeSubspace&&) = default;
+
+  /// Size of the subspace.
+  static constexpr size_t size() { return kSize; }
+  /// Size of the full vector space.
+  static constexpr size_t fullSize() { return kFullSize; }
+
+  /// Axis indices that comprise the subspace.
+  ///
+  /// The specific container and index type should be considered an
+  /// implementation detail. Users should treat the return type as a generic
+  /// container whose elements are convertible to `size_t`.
+  constexpr const std::array<uint8_t, kSize>& indices() const { return m_axes; }
+
+  /// Check if the given axis index in the full space is part of the subspace.
+  constexpr bool contains(size_t index) const {
+    bool isContained = false;
+    for (auto a : m_axes) {
+      isContained |= (a == index);
+    }
+    return isContained;
+  }
+
+  /// Project a full vector into the subspace.
+  ///
+  /// @tparam fullspace_vector_t Veector type in the full space
+  /// @param full Vector in the full space
+  /// @return Subspace vector w/ just the configured axis components
+  ///
+  /// @note Always returns a column vector regardless of the input
+  template <typename fullspace_vector_t>
+  auto projectVector(const Eigen::MatrixBase<fullspace_vector_t>& full) const
```

Comment:
Ideally I think we should try to stick to the projector matrix and do the matrix mult externally. This duplicates that functionality right? If so, I'd argue for this to be removed.